### DISCUSSION TITLE
Looking for testers for lazydev.nvim!

### DISCUSSION THREAD

**Discussion by folke** [LazyVim/LazyVim#3418] [General]
I just create a small new plugin, which is a replacement for `neodev.nvim`.
It improves performance of lua_ls a lot.

For more details, see https://github.com/folke/lazydev.nvim

Let me know what you think!

**Comment by dpetka2001** [comment]
From a first glance it looks really good. It really does load only modules which are required and makes setting `plugins = false` in Neoconf unnecessary and `nvim-cmp` completions are way snappier cuz of less indexing (at least that's what I used to do in my personal config except the LazyVimDev that I use for testing locally forked LazyVim and lazy.nvim repos). Do you happen to know how I can make `LazyVim` global be recognized? I tried setting `library = { vim.env.LAZY .. "/LazyVim" }` but it only recognizes the `config` field in `LazyVim` global variable.

**Reply by folke** [reply]
Yes, completions are much faster for me as well. Huge difference.
I'll look into the `LazyVim` thing!

**Reply by dpetka2001** [reply]
I really like it, because I had dabbled around with a local `.luarc.json` and set up manually `lua_ls` and I had also seen big improvement, because the indexing was way less. But I had to constantly add/remove entries from `.luarc.json` according to what I wanted and to maintain a balance as far as performance was concerned. 

This plugin makes what I did unnecessary, as it will dynamically load modules based on the open files that you have and what are required and still see a big improvement to auto-completing and indexing. 

Really good job @folke. As always thank you a ton!! :heart:

**Reply by folke** [reply]
Thank you!

Just pushed an update that automatically adds `/lua` dirs when needed.

This is what I use:

```lua
  {
    "folke/lazydev.nvim",
    ft = "lua",
    opts = {
      runtime = "~/projects/neovim/runtime",
      library = {
        vim.env.LAZY .. "/luvit-meta/library",
        vim.env.LAZY .. "/wezterm-types",
        "~/projects/LazyVim", -- I use the dev version, so cant do vim.env.LAZY here
      },
      enabled = true,
    },
  },
```

With this `LazyVim` works properly in my dotfiles

**Reply by folke** [reply]
To get the correct path, you could also just make opts a function and then use `require('lazy.core.config').plugins` to get the actual plugin paths

**Reply by folke** [reply]
With this plugin and the `.lazy.lua` file support, I'm also thinking of removing `neoconf.nvim` as a core plugin in LazyVim.
I personally only used neoconf for lsp opts per project (which can now be done with `.lazy.lua`) and for configuring neodev (which is no longer needed with lazydev)

**Reply by dpetka2001** [reply]
Yes, now `LazyVim` provides all fields available compared to before. Thanks!!

And thanks for the info about getting the paths from Lazy directly. Not using it yet, but still useful to know :)

PS: Hope this replaces `neodev` as default and as for Neoconf you know best. I barely used it except for disabling the plugins from indexing and found `.lazy.lua` easier as concept.

**Reply by dpetka2001** [reply]
One more question from my part if you don't mind. Your suggestion above was to use `require("lazy.config.core").plugins`, but I've seen here in LazyVim repo you preferred `require("lazy.config.core").spec.plugins`. Would you mind explaining the difference between the 2?

**Reply by folke** [reply]
Yes, so plugins is the final list of plugins that is set **AFTER** the full spec has been parsed.
While `spec.plugins` is available **DURING** parsing of the main spec.

**Reply by folke** [reply]
So if you want to use it inside an opts functiojn, you can just use plugins. That's always safer than useing spec.plugins (which may change)

**Reply by dpetka2001** [reply]
Cool!! Thank you for the explanation :D

**Reply by folke** [reply]
The reason that `spec.plugins` exists is for users that (in most cases accidentally) `require` modules from plugins during the parsing phase. Lazy still needs to be able to then locate and load the module.

LazyVim of course also highly depends on this, since we do some stuff during parsing as well.

**Reply by dpetka2001** [reply]
So, in `LazyVim.has()` why did you change it for example? It used to be `plugins` and later changed it to `spec.plugins`. From what I can see from a quick search in the repo, most instances of `LazyVim.has()` happen inside opts functions. And with my simple thinking, isn't it safer to check if a plugin exists after the parsing phase has finished and not during it?

**Comment by abeldekat** [comment]
Very nice!

The readme mentions a dependency on lazy.nvim. However, looking at the source code, lazy.nvim is only used in buf.on_require.

I think this approach could also be valuable for configs using other package managers( or even: no package managers). Would it be possible to make the on_require function configurable? As far as I understand the job of that function is to find new paths to complete from.

**Reply by folke** [reply]
You could just do `require('lazydev.buf').on_require = function(...)...end` :)

But it might be hard to do this without the functionality lazy implements for this.

**Reply by folke** [reply]
Actually it's probably not that hard since you have `vim.loader` available.
`Util.get_unloaded_rtp` should return a list of paths to plugins that have not yet been loaded. (like opt plugins from get_nvim_runtime_file)
And in the part below that, just try to strip off anything till `/lua`.

**Reply by abeldekat** [reply]
I think the idea is great! Might also save considerable RAM. I now sometimes have to tweak .luarc.json just decrease the ram usage on my old laptop. 

Perhaps support could be built-in?

I am thinking of an option in the config defaulting to the current "lazy.nvim strategy". Non lazy users could configure that option as ie. "native" explicitly.

**Reply by folke** [reply]
I'm probably going to add support for Neovim packadd. Shouldn't be too hard actuallly.

**Reply by abeldekat** [reply]
Super!

**Reply by abeldekat** [reply]
I would gladly test....)

**Reply by folke** [reply]
Just pushed the update, so would be great if you could test it! :)

**Reply by folke** [reply]
no need to configure anything

**Reply by abeldekat** [reply]
[Plugin](https://github.com/abeldekat/nvim_pde/blob/0cef7da475d3db6c621eeec41c38d9cfdd7f521d/lua/ak/deps/lang/lsp.lua#L17)

[setup](https://github.com/abeldekat/nvim_pde/blob/0cef7da475d3db6c621eeec41c38d9cfdd7f521d/lua/ak/config/lang/lspconfig.lua#L81)

Amazing!

**Reply by folke** [reply]
Great!
Btw, you can load the lazydev plugin anywhere/anytime.
Loading order like with neodev is no longer important.

**Reply by abeldekat** [reply]
I think it's a great improvement. Also happy to see that the lazyrc has been moved to lazy.nvim. Thanks Folke!

**Reply by folke** [reply]
Thank you for the great idea and implementation!
I'm going to remove neoconf as a core plugin in LazyVim soon, since it's not really needed anymore with the lazyrc stuff

**Comment by folke** [comment]
Just added support for `---@module 'foobar'` and a fast cmp completion source for require statements and module annotations.

Check the docs!

**Comment by abeldekat** [comment]
Hi Folke,

One of the side effects of the approach: When browsing modules in the config, the lsp is quite busy. I sometimes see "undefined global require" diagnostic warnings popup temporarily. 
This happens consistently in my config:

1. nvim
2. open lua/ak/config/lang/mason.lua
3. wait for the lsp feedback in jhui to disappear
4. open lua/ak/config/lang/lspconfig.lua
5. observe a flashing "undefined global require", for example on line 9

Most notably, this does not happen when opening lspconfig.lua **directly** after mason.lua...

EDIT: It also does not happen when skipping both step 2 and step 3
EDIT2: It does not seem to be caused by lazy loading in the config. I tested by disabling lazy loading in ak.boot.deps, assigning the "later" function to "now".

**Reply by folke** [reply]
But after loadiding the diagnostics disappear again right? So all good?
Ive also just pushed a change that debounces LuaLS updates, so that should also help.

**Reply by folke** [reply]
That's just LuaLS though. It seems that when a new config is applied, it temporarily removes its entire workspace and returns incorrect diagnostics. But those should automatically disappear once LuaLS is done loading the new config.

**Reply by abeldekat** [reply]
Yes they disappear. I'll try the debounce version.

**Reply by abeldekat** [reply]
I guess the debounce version is very useful in order to avoid unnecessary work. However, it does not prevent the "flashing". Once you'll notice, that flashing becomes rather disturbing.

I attached a video: The "flash" happens around 0:34


https://github.com/LazyVim/LazyVim/assets/58370433/9b6b5afe-04ac-4513-99fb-64ca104ad3ba

**Reply by dpetka2001** [reply]
I also notice this, but I'm not sure I did yesterday. Did you also notice it yesterday with one of the initial releases?

**Reply by abeldekat** [reply]
I think I did. For now, I guess it always happen when browsing and allowing the lsp feedback for each buffer to complete. Never happens on the first buffer.

**Reply by folke** [reply]
As I said, this is more like a LuaLS bug than anything else. Not sure I can do something about it.

**Reply by folke** [reply]
It will also only happen if there are workspace library changes

**Reply by dpetka2001** [reply]
Personally, I don't mind this visual glitch. This is a huge improvement compared to before and even compared to having to maintain a `.luarc.json` yourself.

**Reply by abeldekat** [reply]
Interesting.

Also weird: When, in my scenario, in step 4, I first quickly switch to another file, and immediately after open lspconfig, the flash does not happen.

**Reply by folke** [reply]
Again, that's LuaLS. LuaLS only starts publishing diagnostics once the initial workspace has been loaded, so that's why you don't see it flash initially.

**Reply by abeldekat** [reply]
@folke,

The plugin is working very well with mini.deps. Thanks a lot!

I was wondering what would be the best approach to activate the plugin once(peformaing all initial setup to the library) and then disable. Perhaps the LspNotify callback?
My laptop has very limited ram, so I might add a toggle for vim.g.lazydev_enabled.

Also: Some repositories have a .luarc.json. Would it be best to add a local rc file disabling lazydev? If so, perhaps lazydev could detect this situation automatically.

**Reply by folke** [reply]
The plugin itself uses pretty much no ram. It's LuaLS that will be using your memory.

I could auto-detect `.luarc.json` and don't do lazydev, but that will undoubtly lead to lots of issues from users saying that lazydev doesn't work.

Will add the `.luarc.json` check as an example in the readme. Good point!

**Reply by folke** [reply]
I was able to fix those incorrect diagnostics when reloading a workspace in LuaLS.
https://github.com/LuaLS/lua-language-server/pull/2692

Hopefully it gets merged/released soon!

**Reply by abeldekat** [reply]
That's great! I "upvoted" the pr. Would never have guessed that only 3 lines of code would fix the issue...

**Reply by folke** [reply]
Yeah, but it took me a long time to figure that out 😅

**Reply by abeldekat** [reply]
I admire that persistence!

> The plugin itself uses pretty much no ram. It's LuaLS that will be using your memory.

Indeed. However, `lazydev` incrementally adds plugins to the library of `lua_ls`, thus increasing RAM when browsing config files.
On my old laptop, having 3.2G RAM, I even have to be aware of the amount of `firefox` tabs...
I performed some [measurements](https://github.com/abeldekat/nvim_pde/blob/bed95c8f80a9b868346b3ba9cac7de06748bac8a/lua/ak/config/lang/lazydev.lua#L12), resulting in a worst-case-scenario of 1400M, browsing most of the files in my config. Processors are quite active.

Before `lazydev`, I used a [`.luarc.jsonc`](https://github.com/abeldekat/nvim_pde/blob/bed95c8f80a9b868346b3ba9cac7de06748bac8a/.luarc.jsonc.bak#L23). I added paths by un-commenting relevant lines for plugins in that `.luarc.jsonc`, whenever needed. The `lsp` updates immediately when saving the file. After the session, I would git restore that file. Needless to say, I consider `lazydev` a big improvement.

I [patched](https://github.com/abeldekat/nvim_pde/blob/89cf6615ca44f905b2d1ba26fdf3094cbc51a116/lua/ak/config/lang/lazydev.lua#L27) `lazydev` in my config to achieve the following behavior:

1. On startup, `lazydev` updates its workspace settings once, attaching to the current buffer. It will also perform that action when opening a file in a different workspace in the same Neovim session.
2. For subsequent buffers, `lazydev` does not attach. I have a keymapping to ensure that `lazydev` is attached to the current buffer, if I need the completions.
3. I also have a keymapping to toggle the patch to revert to normal `lazydev` operation.

I would very much welcome your feedback, although I don't expect this functionality to be added to `lazydev`. Writing the patch was a lot of fun. 

On a sidenote, when using native package management, I expect more plugins to be loaded in general. Using lazy.nvim, without the `config` in the `spec`, there will be no `require` to trigger the load.

**Comment by dpetka2001** [comment]
@folke I'm observing some weird behavior with latest update. Screencast following

[Screencast 2024-06-02 16:13:09.webm](https://github.com/LazyVim/LazyVim/assets/12776461/0972960b-b983-44d0-b6fb-855fcc44d90f)

After opening the first buffer and workspace finishes loading, I go to another buffer and after the workspace finishes loading, I go back to the first buffer I opened and now `ipairs` and `table` are `undefined global` as you can see.

**Reply by dpetka2001** [reply]
Testing with different commit hashes, it seems that https://github.com/folke/lazydev.nvim/commit/2e715cd4629df2b47d92468935226655cb7e88ed introduced this and staying in the previous commit doesn't reproduce this.

**Reply by folke** [reply]
Yes, I'm aware of that.
I'm changing it so it can deal with multiple workspace folders.

**Reply by folke** [reply]
I reverted this for now. Do let me know if you'd still see that behavior happening. Thanks!

**Reply by dpetka2001** [reply]
Not anymore. If I notice anything else, I will let you know.

**Reply by folke** [reply]
Just pushed a bunch of changes. Library usage is now tracked per workspace folder.
Let's say your in LazyVim and then go to definition of a Neovim method, then
lspconfig will create a new workspace folder.
Each workspace folder gets the global libraries and just the libs used in the opened files in those workspaces.
So the neovim workspace woudl not get the LAzyVim lib etc

**Reply by dpetka2001** [reply]
Not gonna lie. I don't really understand in depth what you're saying. So, thank you for the example you provided. Can confirm it happens as you say (of course it would :smile: ) and the previous problem didn't surface. So, as just a simple beginner user, I'm just enjoying a better behavior and performance improvement compared to before.
